### PR TITLE
consolidate infrastructure into `pyproject.tml`

### LINKED ISSUE #29: [INFRA] tag `fastpair==0.1.0` causing on issue on install
While working on (3.) in https://github.com/carsonfarmer/fastpair/issues/20, I am running into the problem of there being no associated [*dynamic* version](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#version) with the [`fastpair==0.1.0` tag](https://github.com/carsonfarmer/fastpair/releases/tag/fastpair%3D%3D0.1.0), which is causing the `pip install -e .` to fail.

The two options (IMHO) are a conservative approach [1,2] or a more aggressive approach [2] that does it all at once.

1. conservative approach (2 PRs):
   1. explicitly declaring `version = "v0.1.0"` in `pyproject.toml` in a first PR addressing (3.)
   2. (perhaps create a `v0.1.1` tag and associated release, but this might be overkill)
   3. proceed to step 2
2. one stroke (1 PR):
   1. delete the `fastpair==0.1.0` tag (remote & local)
   2. declare `dynamic = ["version"]` in `pyproject.toml`

@carsonfarmer Let me know if (A) you have any questions/concerns here; and (B) which approach you'd like me to take.

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR:
* consolidate infrastructure into `pyproject.tml`
* resolves #29 
* sets two pins to scipy: `scipy>=1.10,<1.12`
  * the minimum pin follows [SPEC000](https://scientific-python.org/specs/spec-0000/) and we will keep until we bump in the future
  * the maximum pin is temporary until #22 is addressed
* moves to dynamic versioning with `setup tools_scm`, which tracks tags to determine current release and guessed current development version. For example:
  * the current release is `v0.1.0`
  * (locally for me) the current guessed development state is `v0.1.1.dev3+g4ec960e.d20240529`
    * this means...
      * the next patch tag/release would be `v0.1.1`
      * there have been `3` commits since the `v0.1.0` tag
      * the last commit hash ID is `g4ec960e.d20240529`
  * ***after each tag/release developers must reinstall for consistent versioning***
* xref #20 

@carsonfarmer I am not trying to pull a quick one on you or move faster than you are comfortable. Please do not hesitate to raise any questions/concerns if these changes are overwhelming (or you simply need clarification).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+requires = ["setuptools>=61.0", "setuptools_scm[toml]>=6.2"]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools_scm]
+
+[project]
+name = "fastpair"
+dynamic = ["version"]
+authors = [
+    { name = "Carson J. Q. Farmer", email = "carsonfarmer@gmail.com" },
```
Comment (carsonfarmer) [inline pyproject.toml:11]:
```suggestion
    { name = "Carson J. Q. Farmer", email = "carson.farmer@gmail.com" },
```

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+]
+
+#maintainers = [{ name = "" }]
```
Comment (carsonfarmer) [inline pyproject.toml:14]: Interested in adding yourself here?

File: fastpair/__init__.py [hunk 3]
```diff
@@ -10,4 +10,10 @@
 # Copyright (c) 2002-2015, David Eppstein
 # Licensed under the MIT Licence (http://opensource.org/licenses/MIT).
 
+import contextlib
+from importlib.metadata import PackageNotFoundError, version
+
 from .base import FastPair
+
+with contextlib.suppress(PackageNotFoundError):
```
Comment (carsonfarmer) [inline fastpair/__init__.py:18]: Why suppress here?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: MIT License",
+    "Natural Language :: English",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3",
+    "Topic :: Scientific/Engineering :: Information Analysis",
+    "Topic :: System :: Distributed Computing",
+]
+
+requires-python = ">=3.10"
```
Comment (carsonfarmer) [inline pyproject.toml:34]: is this going to be a problem, seems like a high-bar requirement?

**Review by carsonfarmer — APPROVED** [review]
I've checked a buncha this stuff locally. And looked up all the packages. It looks pretty straight-forward. Left a few questions of clarification for you.

File: pyproject.toml (hunk 4 above)
Comment (jGaboardi) [reply pyproject.toml:34]: While we could support Python <3.10 [the SPEC000 support window](https://scientific-python.org/specs/spec-0000/#support-window) is actually closing on Python 3.10 pretty soon. Of course, we don't have to follow this explicitly, but I can say that it has been a life-saver in terms of clear cut minimum requirements.

If you would feel more comfortable in supporting an older minimum version, we can do that, but I wouldn't recommend anything older than Python 3.9.

It's also worth noting that SciPy itself follows SPEC000 and [requires `python>=3.10`](https://github.com/scipy/scipy/blob/3535b3780669d5b742813638461bca79d3358483/pyproject.toml#L47)

File: fastpair/__init__.py (hunk 3 above)
Comment (jGaboardi) [reply fastpair/__init__.py:18]: There is absolutely a reason, and I have of course forgotten exactly what it was 😅 . Let me do some research to provide a concrete justification.

Comment (jGaboardi) [reply fastpair/__init__.py:18]: @carsonfarmer 

So the long-short of it is that `setuptools_scm` is great at being a lightweight, streamlined, & automated VCS when compared to all the needed upkeep with a tool like [`versioneer`](https://github.com/python-versioneer/python-versioneer). Of course there are drawbacks. One is the comment I made when I opened this PR:

> after each tag/release developers must reinstall for consistent versioning

Another is that there is *potential* for mussiness in [conflicting versions](https://stackoverflow.com/a/76206192/8692996) and other edge cases with `pyproject.toml` & `sdist`. Having that safeguard allows for the user to import and use the package even if the (uncommon) mussiness occurs.

Moreover, the reason for not using `try/except` is due to [ruff: SIM105](https://docs.astral.sh/ruff/rules/suppressible-exception/), which we will be enforcing in a subsequent PR.

The pattern we have here:

```python
import contextlib
from importlib.metadata import PackageNotFoundError, version

...

with contextlib.suppress(PackageNotFoundError):
    __version__ = version(<PACKAGE>)
```

is what we have adopted across the PySAL ecosystem (where this modern infrastructure is used – e.g. [`libpysal`](https://github.com/pysal/libpysal/blob/5bda115749e6c8a57323fe86d74487dff3878a4c/libpysal/__init__.py#L24C1-L30C38))

cc @martinfleis

**Comment by jGaboardi** [comment]
> I've checked a buncha this stuff locally. And looked up all the packages. It looks pretty straight-forward. Left a few questions of clarification for you.

Sounds good. I've left some justifications and will wait for your re-review.

File: pyproject.toml (hunk 4 above)
Comment (carsonfarmer) [reply pyproject.toml:34]: Naw, all good. Was more just a question.

**Review by carsonfarmer — APPROVED** [review]

**Pull request merged**